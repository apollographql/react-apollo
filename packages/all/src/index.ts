// Common + hooks
export * from '@apollo/react-common';

// @apollo/react-components
export {
  Query,
  Mutation,
  Subscription,
  QueryComponentOptions,
  MutationComponentOptions,
  SubscriptionComponentOptions
} from '@apollo/react-components';

// @apollo/react-hoc
export {
  graphql,
  withQuery,
  withMutation,
  withSubscription,
  withApollo,
  QueryControls,
  DataValue,
  DataProps,
  MutateProps,
  ChildProps,
  ChildDataProps,
  ChildMutateProps,
  OptionProps,
  OperationOption,
  WithApolloClient
} from '@apollo/react-hoc';

// SSR
export {
  getMarkupFromTree,
  getDataFromTree,
  renderToStringWithData
} from '@apollo/react-ssr';
