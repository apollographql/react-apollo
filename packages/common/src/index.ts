export {
  ApolloContextValue,
  ApolloContextOptions,
  getApolloContext,
  resetApolloContext
} from './context/ApolloContext';
export { ApolloProvider } from './context/ApolloProvider';
export { ApolloConsumer } from './context/ApolloConsumer';

export {
  parser,
  operationName,
  DocumentType,
  IDocumentDefinition
} from './parser/parser';

export * from './types/types';
