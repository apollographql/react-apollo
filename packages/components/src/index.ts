export {
  ApolloProvider,
  ApolloConsumer,
  getApolloContext,
  resetApolloContext
} from '@apollo/react-common';

export { Query } from './Query';
export { Mutation } from './Mutation';
export { Subscription } from './Subscription';

export * from './types';
