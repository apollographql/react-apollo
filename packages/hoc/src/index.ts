export { graphql } from './graphql';

export { withQuery } from './query-hoc';
export { withMutation } from './mutation-hoc';
export { withSubscription } from './subscription-hoc';
export { withApollo } from './withApollo';

export {
  ApolloProvider,
  ApolloConsumer,
  getApolloContext,
  resetApolloContext
} from '@apollo/react-common';

export * from './types';
