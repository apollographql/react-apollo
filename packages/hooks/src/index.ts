export {
  ApolloProvider,
  ApolloConsumer,
  getApolloContext,
  resetApolloContext
} from '@apollo/react-common';

export { useQuery } from './useQuery';
export { useLazyQuery } from './useLazyQuery';
export { useMutation } from './useMutation';
export { useSubscription } from './useSubscription';
export { useApolloClient } from './useApolloClient';

export { RenderPromises } from './ssr/RenderPromises';

export * from './types';
