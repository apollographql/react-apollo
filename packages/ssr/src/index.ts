export { getMarkupFromTree, getDataFromTree } from './getDataFromTree';
export { renderToStringWithData } from './renderToStringWithData';
export * from './types';
