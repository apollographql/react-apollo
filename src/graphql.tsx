import * as React from 'react';
import { DocumentNode } from 'graphql';
import { parser, DocumentType } from './parser';
import { OperationOption, DataProps, MutateProps } from './types';

import { withQuery } from './query-hoc';
import { withMutation } from './mutation-hoc';
import { withSubscription } from './subscription-hoc';

export function graphql<
  TProps extends TGraphQLVariables | {} = {},
  TData = {},
  TGraphQLVariables = {},
  TChildProps = Partial<DataProps<TData, TGraphQLVariables>> &
    Partial<MutateProps<TData, TGraphQLVariables>>
>(
  document: DocumentNode,
  operationOptions: OperationOption<TProps, TData, TGraphQLVariables, TChildProps> = {},
) {
  switch (parser(document).type) {
    case DocumentType.Mutation:
      return withMutation(document, operationOptions);
    case DocumentType.Subscription:
      return withSubscription(document, operationOptions);
    // case DocumentType.Fragment:
    //   throw new Error('fragments cannont currently be used on their own');
    case DocumentType.Query:
    default:
      return withQuery(document, operationOptions);
  }
}
