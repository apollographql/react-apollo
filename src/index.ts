export * from './browser';
export * from './server';
