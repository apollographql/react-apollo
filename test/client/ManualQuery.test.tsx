import * as React from 'react';
import ApolloClient, { NetworkStatus } from 'apollo-client';
import { mount, ReactWrapper } from 'enzyme';
import { InMemoryCache as Cache } from 'apollo-cache-inmemory';
import { ApolloProvider, ManualQuery } from '../../src';
import { MockedProvider, mockSingleLink } from '../../src/test-utils';
import catchAsyncError from '../test-utils/catchAsyncError';
import stripSymbols from '../test-utils/stripSymbols';
import { DocumentNode } from 'graphql';
import gql from 'graphql-tag';

const allPeopleQuery: DocumentNode = gql`
  query people {
    allPeople(first: 1) {
      people {
        name
      }
    }
  }
`;

const allPeopleQueryWithVariables: DocumentNode = gql`
  query people($first: Int!) {
    allPeople(first: $first) {
      people {
        name
      }
    }
  }
`;

interface Data {
  allPeople: {
    people: Array<{ name: string }>;
  };
}

const allPeopleData: Data = {
  allPeople: { people: [{ name: 'Luke Skywalker' }] },
};
const allPeopleMocks = [
  {
    request: { query: allPeopleQuery },
    result: { data: allPeopleData },
  },
];

const allPeopleWithVariablesMocks = [
  {
    request: { query: allPeopleQueryWithVariables, variables: { first: 1 } },
    result: { data: allPeopleData },
  },
];

class AllPeopleQuery extends ManualQuery<Data, { first: number }> {}

it('errors if there is no client instance in the context', () => {
  expect.assertions(1);
  console.error = jest.fn();
  expect(() => mount(<AllPeopleQuery>{() => null}</AllPeopleQuery>)).toThrow(
    'Could not find "client" in the context of ManualQuery. Wrap the root component in an <ApolloProvider>',
  );
});

it('passes a load function as the first param of the render prop', done => {
  let count = 0;
  const Component = () => (
    <AllPeopleQuery query={allPeopleQueryWithVariables} variables={{ first: 1 }}>
      {(load, result) => {
        catchAsyncError(done, () => {
          if (count === 0) {
            expect(result).toEqual({
              loading: false,
              error: undefined,
              data: undefined,
              called: false,
            });

            setTimeout(() => load());
          } else if (count === 1) {
            expect(result).toEqual({
              loading: true,
              error: undefined,
              data: undefined,
              called: true,
            });
          } else if (count === 2) {
            expect(result).toMatchSnapshot();
            done();
            setTimeout(() => load());
          } else if (count === 3) {
            expect(result).toEqual({
              loading: true,
              called: true,
            });
            done();
          }

          count++;
        });

        return null;
      }}
    </AllPeopleQuery>
  );

  mount(
    <MockedProvider mocks={allPeopleWithVariablesMocks} addTypename={false}>
      <Component />
    </MockedProvider>,
  );
});

it('accepts variables in the load function that take precendence over those passed in the props', done => {
  let count = 0;

  const Component = () => (
    <AllPeopleQuery query={allPeopleQueryWithVariables} variables={{ first: 2 }}>
      {(load, result) => {
        catchAsyncError(done, () => {
          if (count === 0) {
            setTimeout(() =>
              load({
                variables: { first: 1 },
              }).then(data => {
                expect(data).toMatchSnapshot('return of load() function');
              }),
            );
          } else if (count === 2) {
            expect(result).toMatchSnapshot('result props');
            done();
          }

          count++;
        });

        return null;
      }}
    </AllPeopleQuery>
  );

  mount(
    <MockedProvider mocks={allPeopleWithVariablesMocks} addTypename={false}>
      <Component />
    </MockedProvider>,
  );
});

it('manages an error', done => {
  let count = 0;

  const Component = () => (
    <AllPeopleQuery query={allPeopleQuery}>
      {(load, result) => {
        catchAsyncError(done, () => {
          if (count === 0) {
            setTimeout(() => load());
          } else if (count === 2) {
            expect(result).toMatchSnapshot('result props');
            setTimeout(() => load());
          } else if (count === 3) {
            expect(result).toEqual({
              loading: true,
              called: true,
            });
            done();
          }

          count++;
        });

        return null;
      }}
    </AllPeopleQuery>
  );

  const mockError = [
    {
      request: { query: allPeopleQuery },
      error: new Error('error'),
    },
  ];

  mount(
    <MockedProvider mocks={mockError} addTypename={false}>
      <Component />
    </MockedProvider>,
  );
});
