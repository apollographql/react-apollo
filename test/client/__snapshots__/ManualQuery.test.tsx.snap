// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts variables in the load function that take precendence over those passed in the props: result props 1`] = `
Object {
  "called": true,
  "data": Object {
    "allPeople": Object {
      "people": Array [
        Object {
          "name": "Luke Skywalker",
          Symbol(id): "$ROOT_QUERY.allPeople({\\"first\\":1}).people.0",
        },
      ],
      Symbol(id): "$ROOT_QUERY.allPeople({\\"first\\":1})",
    },
    Symbol(id): "ROOT_QUERY",
  },
  "error": undefined,
  "loading": false,
}
`;

exports[`accepts variables in the load function that take precendence over those passed in the props: return of load() function 1`] = `
Object {
  "allPeople": Object {
    "people": Array [
      Object {
        "name": "Luke Skywalker",
        Symbol(id): "$ROOT_QUERY.allPeople({\\"first\\":1}).people.0",
      },
    ],
    Symbol(id): "$ROOT_QUERY.allPeople({\\"first\\":1})",
  },
  Symbol(id): "ROOT_QUERY",
}
`;

exports[`manages an error: result props 1`] = `
Object {
  "called": true,
  "data": undefined,
  "error": [Error: Network error: error],
  "loading": false,
}
`;

exports[`passes a load function as the first param of the render prop 1`] = `
Object {
  "called": true,
  "data": Object {
    "allPeople": Object {
      "people": Array [
        Object {
          "name": "Luke Skywalker",
          Symbol(id): "$ROOT_QUERY.allPeople({\\"first\\":1}).people.0",
        },
      ],
      Symbol(id): "$ROOT_QUERY.allPeople({\\"first\\":1})",
    },
    Symbol(id): "ROOT_QUERY",
  },
  "error": undefined,
  "loading": false,
}
`;
